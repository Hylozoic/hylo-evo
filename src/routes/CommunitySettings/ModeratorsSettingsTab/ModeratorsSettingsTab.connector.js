import { connect } from 'react-redux'
import {
  fetchModeratorSuggestions,
  clearModeratorSuggestions,
  getModerators,
  addModerator,
  removeModerator
} from './ModeratorsSettingsTab.store'
import getPerson from 'store/selectors/getPerson'
import { includes } from 'lodash/fp'

export function mapStateToProps (state, props) {
  const moderators = getModerators(state, props)
  const moderatorIds = moderators.map(m => m.id)
  const moderatorSuggestions = state.ModeratorsSettings
    .filter(personId => !includes(personId, moderatorIds))
    .map(personId => getPerson(state, {personId}))

  return {
    moderators,
    moderatorSuggestions
  }
}

export function mapDispatchToProps (dispatch, props) {
  const { communityId } = props

  return {
    addModerator: id => dispatch(addModerator(id, communityId)),
    removeModerator: (id, isRemoveFromCommunity) => dispatch(removeModerator(id, communityId, isRemoveFromCommunity)),
    fetchModeratorSuggestions: autocomplete => dispatch(fetchModeratorSuggestions(communityId, autocomplete)),
    clearModeratorSuggestions: () => dispatch(clearModeratorSuggestions())
  }
}

export default connect(mapStateToProps, mapDispatchToProps)
