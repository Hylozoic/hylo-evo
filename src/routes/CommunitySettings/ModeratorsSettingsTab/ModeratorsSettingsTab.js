import PropTypes from 'prop-types'
import React, { Component } from 'react'
import styles from './ModeratorsSettingsTab.scss'
import Loading from 'components/Loading'
import { KeyControlledItemList } from 'components/KeyControlledList'
import RemovableListItem from 'components/RemovableListItem'
import { isEmpty, get } from 'lodash/fp'
import { getKeyCode, keyMap } from 'util/textInput'

const { array, func, string } = PropTypes

export default class ModeratorsSettingsTab extends Component {
  static propTypes = {
    moderators: array,
    removeModerator: func,
    addModerator: func,
    slug: string
  }

  componentWillUnmount () {
    this.props.clearModeratorSuggestions()
  }

  render () {
    const {
      moderators,
      removeModerator,
      fetchModeratorSuggestions,
      addModerator,
      moderatorSuggestions,
      clearModeratorSuggestions
    } = this.props

    if (!moderators) return <Loading />

    return <div>
      <div>
        {moderators.map(m =>
          <RemovableListItem
            item={m}
            removeModerator={removeModerator}
            confirmMessage={`Are you sure you want to remove ${m.name}'s moderator powers?`}
            key={m.id} />)}
      </div>
      <AddModerator
        fetchModeratorSuggestions={fetchModeratorSuggestions}
        addModerator={addModerator}
        moderatorSuggestions={moderatorSuggestions}
        clearModeratorSuggestions={clearModeratorSuggestions} />
    </div>
  }
}

export class AddModerator extends Component {
  static propTypes = {
    addModerator: func,
    fetchModeratorSuggestions: func
  }

  constructor (props) {
    super(props)
    this.state = {
      adding: false
    }
  }

  render () {
    const { fetchModeratorSuggestions, addModerator, moderatorSuggestions, clearModeratorSuggestions } = this.props

    const { adding } = this.state

    const toggle = () => {
      this.setState({adding: !adding})
    }

    const onInputChange = e => {
      if (e.target.value.length === 0) return clearModeratorSuggestions()
      return fetchModeratorSuggestions(e.target.value)
    }

    const onChoose = choice => {
      addModerator(choice.id)
      clearModeratorSuggestions()
      toggle()
    }

    const chooseCurrentItem = () => {
      if (!this.refs.list) return
      return this.refs.list.handleKeys({
        keyCode: keyMap.ENTER,
        preventDefault: () => {}
      })
    }

    const handleKeys = e => {
      if (getKeyCode(e) === keyMap.ESC) {
        toggle()
        return clearModeratorSuggestions()
      }
      if (!this.refs.list) return
      return this.refs.list.handleKeys(e)
    }

    const listWidth = {width: get('refs.input.clientWidth', this, 0) + 4}

    if (adding) {
      return <div styleName='add-moderator adding'>
        <div styleName='help-text'>Search here for members to grant moderator powers</div>
        <div styleName='input-row'>
          <input styleName='input'
            placeholder='Type...'
            type='text'
            onChange={onInputChange}
            onKeyDown={handleKeys}
            ref='input' />
          <span styleName='cancel-button' onClick={toggle}>Cancel</span>
          <span styleName='add-button' onClick={chooseCurrentItem}>Add</span>
        </div>
        {!isEmpty(moderatorSuggestions) && <div style={listWidth}>
          <KeyControlledItemList
            ref='list'
            items={moderatorSuggestions}
            onChange={onChoose}
            theme={styles} />
        </div>}
      </div>
    } else {
      return <div styleName='add-moderator add-new' onClick={toggle}>
        + Add New
      </div>
    }
  }
}
