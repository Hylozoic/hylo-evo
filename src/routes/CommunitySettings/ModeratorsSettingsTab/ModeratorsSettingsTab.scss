// AddModerator
.add-moderator {
  margin-top: 15px;
}

.adding {
  width: 100%;
}

.help-text {
  composes: caption-lt-lg from 'css/typography.scss';
  margin-bottom: 10px;
}

.input-row {
  display: flex;
  align-items: center;
}

.input {
  flex: 1;
  margin-right: 10px;
  padding: 5px 10px;
  color: $color-rhino;
  &::placeholder {
    color: $color-rhino-60;
  }
}

.cancel-button {
  composes: text-button from 'css/typography.scss';
  color: $color-rhino-60;
  &:hover {
    color: $color-rhino-80;
  }
  margin-right: 10px;
}

.add-button {
  composes: text-button from 'css/typography.scss';
}

.add-new {
  height: 30px;
  line-height: 30px;
  text-align: center;
  border: 1px dashed $color-ghost;
  composes: text-button from 'css/typography.scss'
}

// KeyControlledListTheme

 .items {
   list-style: none;
   border: 1px solid $color-ghost;
   padding: 0;
   li {
     border-top: 1px solid $color-ghost;
     &:first-child {
       border: none;
     }
   }
 }

.item {
  padding: 5px 5px 5px 11px;
  color: $color-caribbean-green;
}

.item-active {
  background-color: $color-athens-gray-dark;
}
