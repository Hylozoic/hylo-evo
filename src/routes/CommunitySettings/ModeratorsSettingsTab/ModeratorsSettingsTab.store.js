import {
  CLEAR_MODERATOR_SUGGESTIONS,
  FETCH_MODERATOR_SUGGESTIONS,
  ADD_MODERATOR,
  REMOVE_MODERATOR
} from 'store/constants'
import { createSelector as ormCreateSelector } from 'redux-orm'
import orm from 'store/models'

export const MODULE_NAME = 'ModeratorsSettingsTab'

const defaultState = []

export default function reducer (state = defaultState, action) {
  const { error, type, payload } = action
  if (error) return state

  switch (type) {
    case FETCH_MODERATOR_SUGGESTIONS:
      return payload.data.community.members.items.map(m => m.id)
    case CLEAR_MODERATOR_SUGGESTIONS:
      return []
    default:
      return state
  }
}

export function fetchModeratorSuggestions (id, autocomplete) {
  return {
    type: FETCH_MODERATOR_SUGGESTIONS,
    graphql: {
      query: `query ($id: ID, $autocomplete: String) {
        community (id: $id) {
          id
          members (first: 10, autocomplete: $autocomplete) {
            hasMore
            items {
              id
              name
              avatarUrl
            }
          }
        }
      }`,
      variables: {
        id, autocomplete
      }
    },
    meta: {
      extractModel: 'Community'
    }
  }
}

export function clearModeratorSuggestions () {
  return {
    type: CLEAR_MODERATOR_SUGGESTIONS
  }
}

export function addModerator (personId, communityId) {
  return {
    type: ADD_MODERATOR,
    graphql: {
      query: `mutation ($personId: ID, $communityId: ID) {
        addModerator(personId: $personId, communityId: $communityId) {
          id
          moderators (first: 100) {
            items {
              id
              name
              avatarUrl
            }
          }
        }
      }`,
      variables: {personId, communityId}
    },
    meta: {
      personId,
      communityId,
      optimistic: true
    }
  }
}

export function removeModerator (personId, communityId, isRemoveFromCommunity) {
  return {
    type: REMOVE_MODERATOR,
    graphql: {
      query: `mutation ($personId: ID, $communityId: ID, $isRemoveFromCommunity: Boolean) {
        removeModerator(personId: $personId, communityId: $communityId, isRemoveFromCommunity: $isRemoveFromCommunity) {
          id
          moderators (first: 100) {
            items {
              id
              name
              avatarUrl
            }
          }
        }
      }`,
      variables: {personId, communityId, isRemoveFromCommunity}
    },
    meta: {
      personId,
      communityId,
      isRemoveFromCommunity,
      optimistic: true
    }
  }
}

// expects props to be of the form {communityId}
export const getModerators = ormCreateSelector(
  orm,
  state => state.orm,
  (state, props) => props.communityId,
  ({ Community }, id) => {
    const community = Community.safeGet({id})
    if (!community) return []
    return community.moderators.toModelArray()
  }
)
