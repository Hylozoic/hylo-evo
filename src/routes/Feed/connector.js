import { connect } from 'react-redux'
import { SAMPLE_FEED_ITEMS } from './sampleData'

function mapStateToProps (state) {
  return {
    feedItems: SAMPLE_FEED_ITEMS
  }
}

function mapDispatchToProps (dispatch, props) {
  return {}
}

export default connect(mapStateToProps, mapDispatchToProps)
