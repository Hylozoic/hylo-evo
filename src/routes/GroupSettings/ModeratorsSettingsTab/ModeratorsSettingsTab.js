import PropTypes from 'prop-types'
import React, { Component, useEffect, useState } from 'react'
import { withTranslation } from 'react-i18next'
import Loading from 'components/Loading'
import KeyControlledItemList from 'components/KeyControlledList/KeyControlledItemList'
import RemovableListItem from 'components/RemovableListItem'
import Icon from 'components/Icon'
import { isEmpty, get, includes } from 'lodash/fp'
import { getKeyCode, keyMap } from 'util/textInput'
import { personUrl } from 'util/navigation'
import SettingsControl from 'components/SettingsControl'
import SettingsSection from '../SettingsSection'
import ModalDialog from 'components/ModalDialog'
import CheckBox from 'components/CheckBox'
import EmojiPicker from 'components/EmojiPicker'
import general from '../GroupSettings.scss' // eslint-disable-line no-unused-vars
import styles from './ModeratorsSettingsTab.scss' // eslint-disable-line no-unused-vars

const { array, func, string, object } = PropTypes

const emptyRole = {
  color: '',
  description: '',
  emoji: '',
  name: '',
  active: ''
}

const validateRole = ({ name, emoji }) => {
  if (name.length < 3) return false
  if (emoji === '') return false
  return true
}

class ModeratorsSettingsTab extends Component {
  static propTypes = {
    addModerator: func,
    addGroupRole: func,
    addRoleToMember: func,
    group: object,
    moderators: array,
    removeModerator: func,
    removeRoleFromMember: func,
    roles: array,
    slug: string,
    updateGroupRole: func
  }

  constructor (props) {
    super(props)
    this.state = this.defaultEditState()
  }

  state = {
    modalVisible: false
  }

  componentDidUpdate (prevProps, prevState) {
    if (!prevProps.group?.groupRoles && this.props.group?.groupRoles) {
      this.setState(this.defaultEditState())
    }
  }

  componentWillUnmount () {
    this.props.clearModeratorSuggestions()
  }

  defaultEditState () {
    const { group } = this.props

    if (!group) return { roles: [], valid: false }

    const roles = group?.groupRoles?.items

    return {
      roles: roles || [],
      error: null,
      modalVisible: false
    }
  }

  submitRemoveModerator = () => {
    this.props.removeModerator(this.state.moderatorToRemove, this.state.isRemoveFromGroup)
  }

  handleAddRole = () => {
    this.setState({
      roles: [...this.state.roles].concat({ ...emptyRole })
    })
  }

  deleteUnsavedRole = (i) => () => {
    const { t } = this.props
    if (window.confirm(t('Are you sure you want to delete this unsaved role/badge?'))) {
      const newRoles = [...this.state.roles]
      newRoles.splice(i, 1)
      this.setState({
        roles: newRoles
      })
    }
  }

  toggleRoleActivation = (i) => () => {
    const { t } = this.props
    const roles = [...this.state.roles]
    const role = roles[i]
    if (window.confirm(`${t('Are you sure you want to ')}${role.active ? t('deactivate') : t('reactivate')} ${t('this role/badge?')}`)) {
      this.props.updateGroupRole({ active: !role.active, groupId: this.props?.group?.id, groupRoleId: role.id }).then((response) => {
        roles[i] = { ...response.payload.data.updateGroupRole }
        this.setState({ roles })
      })
    }
  }

  updateLocalRole = (i) => (key) => (v) => {
    const value = typeof (v.target) !== 'undefined' ? v.target.value : v
    const role = { ...this.state.roles[i] }
    if (role.changed !== true) role.originalState = { ...role }
    role[key] = value
    role.changed = true
    const roles = [...this.state.roles]
    roles[i] = role
    this.setState({ roles })
  }

  saveRole = (i) => () => {
    const { t } = this.props
    const role = { ...this.state.roles[i] }
    if (validateRole(role)) {
      this.props.addGroupRole({ ...role, groupId: this.props?.group?.id }).then((response) => {
        const roles = this.state.roles
        roles[i] = { ...response.payload.data.addGroupRole }
        this.setState({ roles })
      })
    } else {
      window.alert(t('A role must have a valid emoji and name to be saved'))
    }
  }

  resetRole = (i) => () => {
    const role = { ...this.state.roles[i] }
    const roles = [...this.state.roles]
    roles[i] = { ...role.originalState }
    this.setState({ roles })
  }

  updateRole = (i) => () => {
    const { t } = this.props
    const role = { ...this.state.roles[i] }
    if (validateRole(role)) {
      this.props.updateGroupRole({ ...role, groupId: this.props?.group?.id, groupRoleId: role.id }).then((response) => {
        const roles = this.state.roles
        roles[i] = { ...response.payload.data.updateGroupRole }
        this.setState({ roles })
      })
    } else {
      window.alert(t('A role must have a valid emoji and name to be updated'))
    }
  }

  render () {
    const {
      moderators,
      group = {},
      t
    } = this.props

    const {
      modalVisible,
      isRemoveFromGroup,
      roles
    } = this.state

    const unsavedRolePresent = roles.length > 0 ? roles[roles.length - 1]?.active === '' : false
    if (!moderators) return <Loading />

    return (
      <>
        <SettingsSection>
          <h3>
            {group?.moderatorDescriptorPlural || t('Moderators')}
            <div styleName='styles.help-text'>{t('Who has access to group settings and moderation powers')}</div>
          </h3>
          <ModeratorsList key='mList' {...this.props} removeItem={(id) => this.setState({ modalVisible: true, moderatorToRemove: id })} />
          {modalVisible && <ModalDialog
            key='remove-moderator-dialog'
            closeModal={() => this.setState({ modalVisible: false })}
            showModalTitle={false}
            submitButtonAction={this.submitRemoveModerator}
            submitButtonText={this.props.t('Remove')}>
            <div styleName='styles.content'>
              <div styleName='styles.modal-text'>{this.props.t('Are you sure you wish to remove this moderator?')}</div>
              <CheckBox checked={isRemoveFromGroup} label={this.props.t('Remove from group as well')} onChange={value => this.setState({ isRemoveFromGroup: value })} />
            </div>
          </ModalDialog>}
        </SettingsSection>
        <SettingsSection>
          <h3>{t('Other Roles & Badges')}</h3>
          <div styleName='styles.help-text'>{t('Create additional roles or badges for the group')}</div>
          {roles.map((role, i) => (
            <RoleRow
              {...this.props}
              group={group}
              key={i}
              index={i}
              {...role}
              onChange={this.updateLocalRole(i)}
              onSave={this.saveRole(i)}
              onUpdate={this.updateRole(i)}
              onToggleActivation={this.toggleRoleActivation(i)}
              onDelete={this.deleteUnsavedRole(i)}
              onReset={this.resetRole(i)}
            />
          ))}
          {!unsavedRolePresent && (
            <div styleName='styles.add-role' onClick={this.handleAddRole}>
              <h4>{t('Create new role/badge')}</h4>
              <Icon name='Circle-Plus' styleName='styles.new-role' />
            </div>
          )}
        </SettingsSection>
        <br />
      </>
    )
  }
}

export function ModeratorsList ({ moderators, slug, removeItem, fetchModeratorSuggestions, addModerator, moderatorSuggestions, clearModeratorSuggestions }) {
  return (
    <div>
      <div>
        {moderators.map(m =>
          <RemovableListItem
            item={m}
            url={personUrl(m.id, slug)}
            skipConfirm
            removeItem={removeItem}
            key={m.id}
          />)}
      </div>
      <AddModerator
        fetchModeratorSuggestions={fetchModeratorSuggestions}
        addModerator={addModerator}
        moderatorSuggestions={moderatorSuggestions}
        clearModeratorSuggestions={clearModeratorSuggestions}
      />
    </div>
  )
}

export function AddModerator (props) {
  const { t } = useTranslation()
  const { fetchModeratorSuggestions, addModerator, moderatorSuggestions, clearModeratorSuggestions } = props

  const [adding, setAdding] = useState(false)
  const [inputValue, setInputValue] = useState('')

  const inputRef = useRef()
  const listRef = useRef()

  const toggle = () => setAdding(!adding)

  useEffect(() => {
    return () => clearModeratorSuggestions()
  }, [])

  const inputUpdateHandler = e => {
    const value = e.target.value
    if (value === '') {
      clearModeratorSuggestions()
      toggle()
    }

    const chooseCurrentItem = () => {
      if (!this.refs.list) return
      return this.refs.list.handleKeys({
        keyCode: keyMap.ENTER,
        preventDefault: () => {}
      })
    }

    const handleKeys = e => {
      if (getKeyCode(e) === keyMap.ESC) {
        toggle()
        return clearModeratorSuggestions()
      }
      if (!this.refs.list) return
      return this.refs.list.handleKeys(e)
    }

    const listWidth = { width: get('refs.input.clientWidth', this, 0) + 4 }

    if (adding) {
      return (
        <div styleName='styles.adding'>
          <div styleName='styles.help-text'>{t('Search here for members to grant moderator powers')}</div>
          <div styleName='styles.input-row'>
            <input
              styleName='styles.input'
              placeholder={t('Type...')}
              type='text'
              onChange={onInputChange}
              onKeyDown={handleKeys}
              ref='input'
            />
            <span styleName='styles.cancel-button' onClick={toggle}>{t('Cancel')}</span>
            <span styleName='styles.add-button' onClick={chooseCurrentItem}>{t('Add')}</span>
          </div>
          {!isEmpty(moderatorSuggestions) && <div style={listWidth}>
            <KeyControlledItemList
              ref='list'
              items={moderatorSuggestions}
              onChange={onChoose}
              theme={styles}
            />
          </div>}
        </div>
      )
    } else {
      return (
        <div styleName='styles.add-new' id='add-new' onClick={toggle}>
          + {t('Add New')}
        </div>
      )
    }
    setInputValue(e.target.value)
  }

  const onChoose = choice => {
    addModerator(choice.id)
    clearModeratorSuggestions()
    toggle()
  }

  const chooseCurrentItem = () => {
    if (!listRef.current) return
    return listRef.current.handleKeys({
      keyCode: keyMap.ENTER,
      preventDefault: () => {}
    })
  }

  const handleKeys = e => {
    if (getKeyCode(e) === keyMap.ESC) {
      toggle()
      return clearModeratorSuggestions()
    }
    if (!listRef.current) return
    return listRef.current.handleKeys(e)
  }

  const listWidth = { width: inputRef.current?.clientWidth + 4 }

  return adding ? (
    <div styleName='add-moderator adding'>
      <div styleName='help-text'>{t('Search here for members to grant moderator powers')}</div>
      <div styleName='input-row'>
        <input styleName='input'
          placeholder={t('Type...')}
          type='text'
          onChange={e => inputUpdateHandler(e)}
          value={inputValue}
          onKeyDown={handleKeys}
          ref={inputRef} />
        <span className='cancel-button' styleName='cancel-button' onClick={toggle}>{t('Cancel')}</span>
        <span className='add-button' styleName='add-button' onClick={chooseCurrentItem}>{t('Add')}</span>
      </div>
      {!isEmpty(moderatorSuggestions) &&
      <div style={listWidth}>
        <KeyControlledItemList
          ref={listRef}
          items={moderatorSuggestions}
          onChange={onChoose}
          theme={styles}
        />
      </div>
      }
    </div>
  ) : (<div className='add-new' styleName='add-moderator add-new' onClick={toggle}>{t('+ Add New')}</div>)
}

export const AddModerator = withTranslation()(AddModeratorUntranslated)

function RoleRowUntranslated ({
  active,
  addRoleToMember,
  changed,
  clearModeratorSuggestions,
  description,
  emoji,
  fetchModeratorSuggestions,
  fetchMembersForGroupRole,
  group,
  id,
  name,
  onChange,
  onDelete,
  onToggleActivation,
  onReset,
  onSave,
  onUpdate,
  index,
  rawSuggestions = [],
  removeRoleFromMember,
  t
}) {
  const isDraftRole = active === ''
  const inactiveStyle = (!active && !isDraftRole) ? 'styles.inactive' : ''
  return (
    <div styleName={`styles.role-container ${inactiveStyle}`}>
      <div styleName='styles.action-container'>
        {isDraftRole && (<span onClick={onDelete} styleName='styles.action'><Icon name='Trash' /> {t('Delete')}</span>)}
        {!isDraftRole && changed && (<span styleName='styles.action' onClick={onUpdate}><Icon name='Unlock' /> {t('Save')}</span>)}
        {!isDraftRole && changed && (<span styleName='styles.action' onClick={onReset}><Icon name='Back' /> {t('Revert')}</span>)}
        {!isDraftRole && !changed && (<span styleName='styles.action' onClick={onToggleActivation}><Icon name={active ? 'CircleEx' : 'CircleArrow'} /> {active ? t('Deactivate') : t('Reactivate')}</span>)}
      </div>
      <div styleName='styles.role-row'>
        <EmojiPicker forReactions={false} emoji={emoji} handleReaction={onChange('emoji')} className={styles['emoji-picker']} />
        <div styleName='styles.role-stack'>
          <SettingsControl label='Name' controlClass={styles['settings-control']} onChange={onChange('name')} value={name} />
          <SettingsControl label='Description' controlClass={styles['settings-control']} onChange={onChange('description')} value={description} />
        </div>
      </div>
      {
        isDraftRole
          ? (
            <div styleName='styles.role-row styles.reverse-flex'>
              <div styleName='styles.create-button' onClick={onSave}>{t('Create Role')}</div>
            </div>
          )
          : active && (
            <SettingsSection>
              <GroupRoleList
                {...{ addRoleToMember, rawSuggestions, clearModeratorSuggestions, fetchMembersForGroupRole, fetchModeratorSuggestions, removeRoleFromMember }}
                key='grList'
                groupRoleId={id}
                slug={group.slug}
              />
            </SettingsSection>
          )
      }
    </div>
  )
}

const RoleRow = withTranslation()(RoleRowUntranslated)

class AddMemberToRoleUntranslated extends Component {
  static propTypes = {
    addRoleToMember: func,
    clearSuggestions: func,
    fetchSuggestions: func,
    groupRoleId: string,
    memberSuggestions: array,
    updateLocalMembersForRole: func
  }

  constructor (props) {
    super(props)
    this.state = {
      adding: false
    }
  }

  render () {
    const { fetchSuggestions, addRoleToMember, memberSuggestions, clearSuggestions, groupRoleId, updateLocalMembersForRole, t } = this.props

    const { adding } = this.state

    const toggle = () => {
      clearSuggestions()
      this.setState({ adding: !adding })
    }

    const onInputChange = e => {
      if (e.target.value.length === 0) return clearSuggestions()
      return fetchSuggestions(e.target.value)
    }

    const onChoose = choice => {
      addRoleToMember({ personId: choice.id, groupRoleId }).then(() => {
        updateLocalMembersForRole(choice)
      })
      toggle()
    }

    const chooseCurrentItem = () => {
      if (!this.refs.list) return
      return this.refs.list.handleKeys({
        keyCode: keyMap.ENTER,
        preventDefault: () => {}
      })
    }

    const handleKeys = e => {
      if (getKeyCode(e) === keyMap.ESC) {
        toggle()
        return
      }
      if (!this.refs.list) return
      return this.refs.list.handleKeys(e)
    }

    const listWidth = { width: get('refs.input.clientWidth', this, 0) + 4 }

    if (adding) {
      return (
        <div styleName='styles.adding'>
          <div styleName='styles.help-text'>{t('Search here for members to grant this role too')}</div>
          <div styleName='styles.input-row'>
            <input
              styleName='styles.input'
              placeholder='Type...'
              type='text'
              onChange={onInputChange}
              onKeyDown={handleKeys}
              ref='input'
            />
            <span styleName='styles.cancel-button' onClick={toggle}>{t('Cancel')}</span>
            <span styleName='styles.add-button' onClick={chooseCurrentItem}>{t('Add')}</span>
          </div>
          {!isEmpty(memberSuggestions) && <div style={listWidth}>
            <KeyControlledItemList
              ref='list'
              items={memberSuggestions}
              onChange={onChoose}
              theme={styles}
            />
          </div>}
        </div>
      )
    } else {
      return (
        <div styleName='styles.add-new' onClick={toggle}>
          + {t('Add Member to Role')}
        </div>
      )
    }
  }
}

const AddMemberToRole = withTranslation()(AddMemberToRoleUntranslated)

export function GroupRoleList ({ slug, removeItem, fetchModeratorSuggestions, addRoleToMember, rawSuggestions, clearModeratorSuggestions, groupRoleId, fetchMembersForGroupRole, removeRoleFromMember }) {
  const [membersForRole, setMembersForRole] = useState([])

  useEffect(() => {
    fetchMembersForGroupRole({ groupRoleId })
      .then((response) => {
        setMembersForRole(response.payload.data.group.members.items)
      })
  }, [])

  const memberRoleIds = membersForRole.map(mr => mr.id)

  const memberSuggestions = rawSuggestions.filter(person => !includes(person.id, memberRoleIds))

  const updateLocalMembersForRole = (choice) => {
    const updatedMembers = [...membersForRole, choice]
    setMembersForRole(updatedMembers)
  }

  const handleRemoveRoleFromMember = (id) => {
    removeRoleFromMember({ personId: id, groupRoleId }).then(() => {
      const updatedMembers = membersForRole.filter(member => member.id !== id)
      setMembersForRole(updatedMembers)
    })
  }

  return (
    <div>
      <div>
        {membersForRole.map(m =>
          <RemovableListItem
            item={m}
            url={personUrl(m.id, slug)}
            removeItem={handleRemoveRoleFromMember}
            key={m.id}
          />)}
      </div>
      <AddMemberToRole
        fetchSuggestions={fetchModeratorSuggestions}
        addRoleToMember={addRoleToMember}
        memberSuggestions={memberSuggestions}
        clearSuggestions={clearModeratorSuggestions}
        updateLocalMembersForRole={updateLocalMembersForRole}
        groupRoleId={groupRoleId}
      />
    </div>
  )
}

export default withTranslation()(ModeratorsSettingsTab)
