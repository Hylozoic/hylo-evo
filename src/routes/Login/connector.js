import { connect } from 'react-redux'
import { checkLogin, login } from './actions'
export default connect(null, {checkLogin, login})
