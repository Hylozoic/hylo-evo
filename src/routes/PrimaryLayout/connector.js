import { connect } from 'react-redux'
import { SAMPLE_COMMUNITY } from 'routes/Feed/sampleData'
import { fetchCurrentUser } from './actions'

function mapStateToProps ({currentUser}) {
  return {
    community: SAMPLE_COMMUNITY,
    currentUser
  }
}

export default connect(mapStateToProps, {fetchCurrentUser})
