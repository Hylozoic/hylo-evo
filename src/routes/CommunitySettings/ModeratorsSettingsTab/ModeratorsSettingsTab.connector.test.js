import { mapStateToProps, mapDispatchToProps } from './ModeratorsSettingsTab.connector'
import orm from 'store/models'

let state
beforeAll(() => {
  const session = orm.session(orm.getEmptyState())
  const community = session.Community.create({id: '99', slug: 'foo'})
  session.Community.create({id: '88', slug: 'bar'})

  session.Me.create({
    id: '1',
    memberships: [session.Membership.create({
      id: '345',
      community: community.id,
      hasModeratorRole: true
    })]
  })

  session.Person.create({
    id: '1'
  })

  session.Person.create({
    id: '2'
  })

  session.Membership.create({
    id: '355',
    community: community.id,
    person: '2',
    hasModeratorRole: true
  })

  community.update({moderators: ['1', '2']})

  state = {
    orm: session.state,
    ModeratorsSettings: []
  }
})

describe('mapStateToProps', () => {
  it('works', () => {
    const props = {
      communityId: '99'
    }

    expect(mapStateToProps(state, props)).toMatchSnapshot()
  })
})

describe('mapDispatchToProps', () => {
  it('correctly sets up dispatches', () => {
    const dispatch = jest.fn(x => x)
    const props = {
      communityId: 11
    }

    const dispatchProps = mapDispatchToProps(dispatch, props)
    expect(dispatchProps.addModerator(10)).toMatchSnapshot()
    expect(dispatchProps.removeModerator(10, true)).toMatchSnapshot()
    expect(dispatchProps.fetchModeratorSuggestions('autocomplete')).toMatchSnapshot()
    expect(dispatchProps.clearModeratorSuggestions()).toMatchSnapshot()
  })
})
