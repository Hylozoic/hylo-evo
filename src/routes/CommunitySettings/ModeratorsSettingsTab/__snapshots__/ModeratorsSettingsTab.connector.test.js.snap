// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`mapDispatchToProps correctly sets up dispatches 1`] = `
Object {
  "graphql": Object {
    "query": "mutation ($personId: ID, $communityId: ID) {
        addModerator(personId: $personId, communityId: $communityId) {
          id
          moderators (first: 100) {
            items {
              id
              name
              avatarUrl
            }
          }
        }
      }",
    "variables": Object {
      "communityId": 11,
      "personId": 10,
    },
  },
  "meta": Object {
    "communityId": 11,
    "optimistic": true,
    "personId": 10,
  },
  "type": "ADD_MODERATOR",
}
`;

exports[`mapDispatchToProps correctly sets up dispatches 2`] = `
Object {
  "graphql": Object {
    "query": "mutation ($personId: ID, $communityId: ID, $isRemoveFromCommunity: Boolean) {
        removeModerator(personId: $personId, communityId: $communityId, isRemoveFromCommunity: $isRemoveFromCommunity) {
          id
          moderators (first: 100) {
            items {
              id
              name
              avatarUrl
            }
          }
        }
      }",
    "variables": Object {
      "communityId": 11,
      "isRemoveFromCommunity": true,
      "personId": 10,
    },
  },
  "meta": Object {
    "communityId": 11,
    "isRemoveFromCommunity": true,
    "optimistic": true,
    "personId": 10,
  },
  "type": "REMOVE_MODERATOR",
}
`;

exports[`mapDispatchToProps correctly sets up dispatches 3`] = `
Object {
  "graphql": Object {
    "query": "query ($id: ID, $autocomplete: String) {
        community (id: $id) {
          id
          members (first: 10, autocomplete: $autocomplete) {
            hasMore
            items {
              id
              name
              avatarUrl
            }
          }
        }
      }",
    "variables": Object {
      "autocomplete": "autocomplete",
      "id": 11,
    },
  },
  "meta": Object {
    "extractModel": "Community",
  },
  "type": "FETCH_MODERATOR_SUGGESTIONS",
}
`;

exports[`mapDispatchToProps correctly sets up dispatches 4`] = `
Object {
  "type": "CLEAR_MODERATOR_SUGGESTIONS",
}
`;

exports[`mapStateToProps works 1`] = `
Object {
  "moderatorSuggestions": Array [],
  "moderators": Array [
    SessionBoundModel {
      "_fields": Object {
        "id": "1",
        "skills": undefined,
      },
    },
    SessionBoundModel {
      "_fields": Object {
        "id": "2",
        "skills": undefined,
      },
    },
  ],
}
`;
